### ISSUE TITLE
Hold/unhold turning off call recording and not finishing current recording

### ISSUE THREAD

**Issue by dbmikus** [onsip/rtpengine#10]
# Reproduce Problem

1. Call between two users, with RTP proxied over rtpengine and "record-call=yes" sent in the offer message
2. Place one call on hold
3. Unhold the call

This generates another offer, but this offer might not include the "record-call=yes" flag. This will disable call recording part-way through the call, but we aren't properly freeing memory, closing file descriptors, and moving metadata files to their trigger location when this happens.

# Solutions

## Option A

The user can turn on and off recording during the call at will, through rtpengine messages. If the user sends another message, we can either treat the absence of the "record-call" flag as meaning stop recording or only stop recording if we get a "record-call=no" flag. I think we should only stop recording if we get an explicit "no".

We could also let the user turn back on call-recording stuff, too.

## Option B

Call recording is turned on at the start of a call, or not at all. We will ignore future "record-call" messages.

# My Opinion

I greatly prefer **Option A** where we only turn off recording if another rtpengine message includes "record-call=no". I think, if we do this, we should also let the user turn back on call recording. Everything will still go to the same PCAP file, but we will only write rtpengine when a `record_call` variable is true.

Thoughts?

**Comment by egreenmachine** [comment]
Option A feels right to me

**Comment by egreenmachine** [comment]
We will need to document it either way

**Comment by josephfrazier** [comment]
So, I'm going to be contrarian here, but I might be due to a relative lack of personal experience with the code, so take the following with a grain of salt.

> This generates another offer, but this offer might not include the "record-call=yes" flag.

Assuming the rtpengine user's intent is to continue recording the call, aren't they responsible for sending the "record-call=yes" flag? That is, shouldn't we fix our OpenSIPS?

> Option A: The user can turn on and off recording during the call at will, through rtpengine messages

This would be the [`start recording`](https://github.com/onsip/rtpengine#start-recording-message) message mentioned in the README, accompanied by a `stop recording` message? Or would recording only be toggle-able on a new offer/answer?

> Option B

This seems to require much less work than Option A, and also sounds more in line with our original plans for the interface (before we started allowing the original answer flags to overwrite the original offer flags). I think we should go with this, but perhaps I'm misunderstanding the approaches.

**Comment by etamme** [comment]
We do specifically handle media relaying in the core for sequential requests - it is passed to the route media relay request block with the parameter "sequential" - and we do some special processing as a result.  

I belive that we should more or less do option A, which is to, i believe, fix the core logic so that if we have been recording a call, we continue to record it - aka pass the same call recording flags in the offer/answer.  We do stuff like this now to continue using the specific proxy (rtpproxy, mediaproxy, rtpengine) that we have been - i see this as a similar issue.

I do not see this as a way for users to turn on or off call recording - this is simply a fix to the logic of the core to properly handle sequential requests.

**Comment by dbmikus** [comment]
The difference between A and B is not large for the actual rtpengine code. Option A could be more confusing for our post-processor, but we're going to have to handle gaps in rtp either way if we record calls that can be held. I think Option B is only slightly easier, and it is much more limiting.

The `start recording` message differs from our flag sent over an `offer/answer` message. We need to have a "record-call" flag over the `offer` so that we can pre-empt kernelization. Our recording method can only be toggled by `offer/answer`. Now, if recording is already started on an `offer` message,  there's nothing stopping us from using the `start recording` and `stop recording` messages for that call, but we are currently entirely separate from that rtpengine message.

For your first part:

> Assuming the rtpengine user's intent is to continue recording the call, aren't they responsible for sending the "record-call=yes" flag? That is, shouldn't we fix our OpenSIPS?

I can see what you're saying and it is debatable, but I disagree. The hold/unhold operation, all other things being equal in OpenSIPS (or whatever other routing system), should not have the side-effect of turning off call recording. We have recording set up per call, not per dialogue established by an offer/answer. Thus, a new offer/answer on the same call should not alter the recording state.

I might change my opinion if a call transfer does not start a new call. Do we know if that is the case? I think a call transfer is close enough to an entirely new call that recording rules should not follow it be default, so if a transfer would still be on the same call struct, it should turn recording off.

I'm not really looking at this issue from a configuration standpoint, but rather from a user perspective. I don't think that hold/unhold should turn off recording by default, so I don't think our software should do that either. That being said, I think it would still be worth it to manually send "record-call=yes" flags on the unhold offer through OpenSIPS.

**Comment by dbmikus** [comment]
I think it makes sense to have the software behavior default to the same behavior a user would expect, as long as that doesn't break logical code consistency.

Summary of questions:

1. Should hold/unhold alter recording state by default?
2. Should adding someone to a call alter recording state by default?
3. Should transferring a call alter recording state by default?

I feel clearly about the first question, but not as much so for the other two. My vote for hold/unhold is that it should **not** change the recording state.

The other questions come into some privacy issues. I think most call recording systems warn callers that they may be recorded. Someone being added to a call might miss the initial warning. Should call recording default to off just to be safe?

Transferring a call seems like a new call, and I lean towards saying that call recording should default to off. If transferring a call does indeed make a new call, then I definitely believe that call recording status should default to off.

The questions about introducing a new person to a conversation are related. We might make 2. and 3. behave the same way, or we might delineate their behavior based on whether the new person represents a new SIP call or just SDP renegotiation.

(Note that I'm probably misusing terms. My RFC knowledge is not as strong as some others)

**Comment by josephfrazier** [comment]
@dbmikus, I think we're talking past each other a bit. While it's important for us to figure out the end user's expected behavior (w.r.t. recording) for hold/unhold/monitor/whisper/barge/transfer/etc, I think this issue should stay focused on rtpengine's behavior regarding rtpengine_offer and rtpengine_answer (and `start recording`, `stop recording`, should those come into play).

I agree with @etamme in https://github.com/onsip/rtpengine/issues/10#issuecomment-166953353. It sounds like rtpengine (in its [current state](https://github.com/onsip/rtpengine/commit/332254313588d42691eb72e627bca465e3140724)) will behave as expected if it receives "record-call=yes" with every offer/answer.

**Comment by etamme** [comment]
yeah i basically want to pass whatever flag we need to to make rtpengine keep recording the call - if that means passing no record-call flag, then we can do that, or if we need record-call=true we can do that instead.

**Comment by dbmikus** [comment]
If we do want the default behavior without a "call-recording" flag to stop recording, then we should remove the "=yes" part of the flag, and we just check whether "call-recording" is present or absent in the message.

This does make the behavior dead-simple in the code, even if it strays from expected user behavior.

- - -

@josephfrazier I'm pretty sure I understand, but I'm just getting a little philosophical about code :smiley: 

I think that software code and configuration code should use defaults and behavior that match expected user behavior. I don't think we should have divergent defaults for the user and for configuration.

**Comment by dbmikus** [comment]
And actually, Joseph, if you still disagree, I'm fine with making every offer/answer turn recording off unless we tell the rtpengine to turn it on. But I still believe in my points above :neckbeard:

**Comment by josephfrazier** [comment]
I'm a little unclear on how the "hold" `rtpengine_offer` is handled by rtpengine. rtpengine is aware that this offer/answer is intended to update the previous offer/answer in the logical "call", right? If so, it seems the most user-friendly for the values of the `record-call` flag to behave as follows:

* "yes" -> turn recording on
* "no" -> turn recording off
* other/missing -> keep recording the way it was before

**Comment by dbmikus** [comment]
@josephfrazier Your bullet list was the behavior I advocated. The way the hold is handled goes as follows:

1. Prior to hold, offer comes in with "record-call=yes". In the code that handles offer/answer for a call, we set up recording code
2. Respond with answer. For this, the code still requires "record-call=yes". This should follow whatever behavior we choose from the debate above
3. People talk
4. Alice places Bob on hold. I believe this generates another SDP offer/answer. This offer/answer doesn't have "record-call" set. This frees the recording info, but we don't actually close and move the medata file
5. Alice unholds Bob. Same thing happens as in the previous step.

**Comment by josephfrazier** [comment]
Ok, it sounds like we're in agreement, then. I just wanted to make sure I understood the extent of rtpengine's knowledge about the "call", which persists over multiple offer/answer exchanges (with one per hold/unhold). @egreenmachine, @etamme: do the previous two comments sound good to you?

**Comment by egreenmachine** [comment]
The last 2 comments were the only ones I read since posting last! I think you guys have this under control.

**Comment by dbmikus** [comment]
See https://github.com/onsip/rtpengine/commit/66f5678df57e5d02c07700af10f487501a3e5f16 for the code change.
